lib/eventFactory.js

```diff
@@ -2,9 +2,94 @@
 
 const BaseFactory = require('./baseFactory');
 const Event = require('./event');
+const jobConfigSchema = require('screwdriver-data-schema').config.job;
+const joi = require('joi');
 
 let instance;
 
+/**
+ * Create builds associated with this event
+ * For example, if startFrom ~commit, then create builds with jobs that have requires: ~commit
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {String}   [config.prNum]             PR Number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               Making optional for backward compatibility
+ * @param  {Number}   eventId                    Event's id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    const createJob = (job, jobName) => jobFactory.create({
+        pipelineId: job.pipelineId,
+        name: jobName,
+        permutations: job.permutations
+    });
+
+    return pipeline.jobs.then((jobs) => {
+        const jobNamesArray = jobs.map(j => j.name);
+        const jobsToCreate = [];
+        let jobsToStart = [];
+
+        jobs.forEach((job) => {
+            const enabled = job.state === 'ENABLED';
+            const requires = job.permutations.requires;
+            const isJobName = joi.validate(startFrom, jobConfigSchema.jobname);
+            const isTrigger = joi.validate(startFrom, jobConfigSchema.trigger);
+
+            if (enabled) {
+                // if startFrom is a jobname, then start that job
+                if (!isJobName.error && startFrom === job.name) {
+                    jobsToStart.push(job);
+                // if startFrom is a trigger (example: ~commit, ~pr) then start jobs that include that key
+                } else if (!isTrigger.error && requires && requires.includes(startFrom)) {
+                    const prJobName = `PR-${config.prNum}-${job.name}`;
+                    const prJobIndex = jobNamesArray.indexOf(prJobName);
```

Comment:
what results in this case ? Rerunning PR ?

Reply:
Yes. Two cases can happen: 
- The first time PR is created, no jobs are created yet, so we need to create the job. 
- Rerunning PR or a new commit to push to the PR, then the PR job already exists

Reply:
Found a bug in the "rerunning case" actually. 
Two cases:
- `PR-1-component` doesn't exist yet, `prJobIndex<0`: it creates the job `PR-1-component` first, then concat the newly created job to `jobToStart`, then start the job (create a build of that job). **This case works fine.** 
- `PR-1-component` already exists: let's say the pipeline has 2 jobs [component, PR-1-component]. When job is `component`, prJobName = `PR-1-component`, since prJobIndex >= 0, it start the job, which is `component`, instead of `PR-1-component` (line 68)